File: cpp/graphs/zero_one_bfs_production.cpp
#include <algorithm>
#include <deque>
#include <stdexcept>
#include <utility>
#include <vector>

using Cell = std::pair<int, int>;

std::vector<Cell> zeroOneBfsPath(
    int width,
    int height,
    const std::vector<std::vector<bool>>& walls,
    const std::vector<std::vector<int>>& weights,
    Cell start,
    Cell target
) {
    if (width <= 0 || height <= 0) {
        throw std::invalid_argument("grid dimensions must be positive");
    }

    auto inside = [width, height](Cell cell) {
        return cell.first >= 0 && cell.first < height
            && cell.second >= 0 && cell.second < width;
    };

    if (!inside(start) || !inside(target)) {
        throw std::out_of_range("start and target must be inside the grid");
    }

    if (walls[start.first][start.second] || walls[target.first][target.second]) {
        return {};
    }

    const int total = width * height;
    const int infinity = total + 1;
    std::vector<int> distance(total, infinity);
    std::vector<int> parent(total, -1);
    std::deque<int> deque;

    auto id = [width](Cell cell) {
        return cell.first * width + cell.second;
    };

    const int startId = id(start);
    const int targetId = id(target);
    distance[startId] = 0;
    deque.push_back(startId);

    while (!deque.empty()) {
        const int currentId = deque.front();
        deque.pop_front();
        const Cell current{currentId / width, currentId % width};

        if (currentId == targetId) {
            std::vector<Cell> path;
            for (int stepId = targetId; stepId != -1; stepId = parent[stepId]) {
                path.push_back({stepId / width, stepId % width});
            }
            std::reverse(path.begin(), path.end());
            return path;
        }

        const std::vector<Cell> neighbors = {
            {current.first - 1, current.second},
            {current.first, current.second + 1},
            {current.first + 1, current.second},
            {current.first, current.second - 1},
        };

        for (Cell next : neighbors) {
            if (!inside(next) || walls[next.first][next.second]) {
                continue;
            }

            const int weight = weights[next.first][next.second];
            if (weight != 0 && weight != 1) {
                throw std::invalid_argument("edge weights must be 0 or 1");
            }

            const int nextId = id(next);
            const int nextDistance = distance[currentId] + weight;
            if (nextDistance >= distance[nextId]) {
                continue;
            }

            distance[nextId] = nextDistance;
            parent[nextId] = currentId;
            if (weight == 0) {
                deque.push_front(nextId);
            } else {
                deque.push_back(nextId);
            }
        }
    }

    return {};
}

File: python/graphs/zero_one_bfs_production.py
from collections import deque
from collections.abc import Mapping, Set


Cell = tuple[int, int]


def zero_one_bfs_path(
    width: int,
    height: int,
    walls: Set[Cell],
    weights: Mapping[Cell, int],
    start: Cell,
    target: Cell,
) -> list[Cell]:
    if width <= 0 or height <= 0:
        raise ValueError("grid dimensions must be positive")

    def cell_id(cell: Cell) -> int:
        return cell[0] * width + cell[1]

    def inside(cell: Cell) -> bool:
        row, col = cell
        return 0 <= row < height and 0 <= col < width

    if not inside(start) or not inside(target):
        raise ValueError("start and target must be inside the grid")

    total = width * height
    blocked = bytearray(total)
    edge_cost = bytearray(total)

    for wall in walls:
        if inside(wall):
            blocked[cell_id(wall)] = 1

    for cell, weight in weights.items():
        if not inside(cell) or weight not in (0, 1):
            raise ValueError("weights must be 0 or 1 inside the grid")
        edge_cost[cell_id(cell)] = weight

    start_id = cell_id(start)
    target_id = cell_id(target)
    if blocked[start_id] or blocked[target_id]:
        return []

    infinity = total + 1
    distances = [infinity] * total
    parent = [-1] * total
    queue = deque([start_id])
    distances[start_id] = 0

    while queue:
        current_id = queue.popleft()
        if current_id == target_id:
            path = []
            while current_id != -1:
                path.append(divmod(current_id, width))
                current_id = parent[current_id]
            path.reverse()
            return path

        row, col = divmod(current_id, width)
        neighbors = (
            (row - 1, col),
            (row, col + 1),
            (row + 1, col),
            (row, col - 1),
        )

        for next_row, next_col in neighbors:
            if not (0 <= next_row < height and 0 <= next_col < width):
                continue

            next_id = next_row * width + next_col
            if blocked[next_id]:
                continue

            distance = distances[current_id] + edge_cost[next_id]
            if distance >= distances[next_id]:
                continue

            distances[next_id] = distance
            parent[next_id] = current_id
            if edge_cost[next_id] == 0:
                queue.appendleft(next_id)
            else:
                queue.append(next_id)

    return []
